components/storage/blockdevice/COMPONENT_OSPIF/OSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,1885 @@
+        return -1;
+    }
+
+    // Support 4-Byte address instructions
+    if (four_byte_inst_table[OSPIF_4_BYTE_INST_TABLE_FAST_READ_SUPPORT_BYTE] & FOURBYTE_INST_1_1_1_READ_BITMASK) {
+        _support_4_byte_inst = true;
+    }
+
+    // Detect erase types 4-Byte instructions
+    if (sfdp_info.fbatbl.size > OSPIF_4_BYTE_INST_TABLE_INST_ERASE_TYPE_1_BYTE) {
+        // Loop Erase Types 1-4
+        for (int i_ind = 0; i_ind < 4; i_ind++) {
+            sfdp_info.fbatbl.erase_type_4_byte_inst_arr[i_ind] = OSPI_NO_INST; // Default for unsupported type
+            if (four_byte_inst_table[OSPIF_4_BYTE_INST_TABLE_ERASE_SUPPORT_BYTE] & (FOURBYTE_INST_ERASE_TYPE_1_BITMASK << i_ind)) {
+                sfdp_info.fbatbl.erase_type_4_byte_inst_arr[i_ind] = four_byte_inst_table[OSPIF_4_BYTE_INST_TABLE_INST_ERASE_TYPE_1_BYTE + i_ind];
+            }
+            tr_debug("Erase Type %d - Inst: 0x%xh", (i_ind + 1), sfdp_info.fbatbl.erase_type_4_byte_inst_arr[i_ind]);
+        }
+    }
+
+    return 0;
+}
+
+int OSPIFBlockDevice::_handle_vendor_quirks()
+{
+    uint8_t vendor_device_ids[OSPI_RDID_DATA_LENGTH] = {0};
+    /* Read Manufacturer ID (1byte), and Device ID (2bytes) */
+    ospi_status_t status = _ospi_send_general_command(OSPIF_INST_RDID, OSPI_NO_ADDRESS_COMMAND,
+                                                      NULL, 0,
+                                                      (char *) vendor_device_ids, OSPI_RDID_DATA_LENGTH);
+    if (OSPI_STATUS_OK != status) {
+        tr_error("Read Vendor ID Failed");
+        return -1;
+    }
+
+    tr_debug("Vendor device ID = 0x%x 0x%x 0x%x", vendor_device_ids[0], vendor_device_ids[1], vendor_device_ids[2]);
+
+    switch (vendor_device_ids[0]) {
+        case 0xbf:
+            // SST devices come preset with block protection
+            // enabled for some regions, issue global protection unlock to clear
+            tr_debug("Applying quirks for SST");
+            _clear_protection_method = OSPIF_BP_ULBPR;
+            break;
+        case 0xc2:
+            // Macronix devices have several quirks:
+            // 1. Have one status register and 2 config registers, with a nonstandard instruction for reading the config registers
+            // 2. Require setting a "fast mode" bit in config register 2 to operate at higher clock rates
+            // 3. Should never attempt to enable 4-byte addressing (it causes reads and writes to fail)
+            tr_debug("Applying quirks for macronix");
+            _needs_fast_mode = true;
+            _num_status_registers = 3;
+            _read_status_reg_2_inst = OSPIF_INST_RDCR;
+            //The sfdp parameter values in Macronix ocata flash(memory type ID is 0x85) are all 0xFF,
+            // so we need to define the parameter values by software to support SFDP parsing.
+            if (vendor_device_ids[1] == 0x85) {
+                _need_define_sfdp_para = true;
+                _sfdp_head_table[0] = 0x53;
```

Comment:
@SeppoTakalo Should memory required by a single device(flash module in this case) for a workaround be taken from heap instead of stack?